daemon/emer-circular-file.c

```diff
@@ -0,0 +1,825 @@
+ *
+ * eos-event-recorder-daemon is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * eos-event-recorder-daemon is distributed in the hope that it will be
+ * useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+ * Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with eos-event-recorder-daemon.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include "emer-circular-file.h"
+
+#include <gio/gio.h>
+#include <glib.h>
+
+#include "shared/metrics-util.h"
+
+#define METADATA_GROUP_NAME "metadata"
+#define MAX_SIZE_KEY "max_size"
+#define SIZE_KEY "size"
+#define HEAD_KEY "head"
+
+typedef struct _EmerCircularFilePrivate
+{
+  GFile *data_file;
+  GKeyFile *metadata_key_file;
+
+  GByteArray *write_buffer;
+
+  guint64 max_size;
+  guint64 size;
+  goffset head;
+} EmerCircularFilePrivate;
+
+static void emer_circular_file_initable_init (GInitableIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (EmerCircularFile, emer_circular_file, G_TYPE_OBJECT,
+                         G_ADD_PRIVATE (EmerCircularFile)
+                         G_IMPLEMENT_INTERFACE (G_TYPE_INITABLE, emer_circular_file_initable_init))
+
+enum
+{
+  PROP_0,
+  PROP_PATH,
+  PROP_MAX_SIZE,
+  NPROPS
+};
+
+static GParamSpec *emer_circular_file_props[NPROPS] = { NULL, };
+
+static gchar *
+get_metadata_filepath (EmerCircularFile *self)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  gchar *data_filepath = g_file_get_path (priv->data_file);
+  gchar *metadata_filepath =
+    g_strconcat (data_filepath, METADATA_EXTENSION, NULL);
+  g_free (data_filepath);
+
+  return metadata_filepath;
+}
+
+static gboolean
+save_metadata_file (EmerCircularFile *self,
+                    GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  gchar *metadata_filepath = get_metadata_filepath (self);
+  gboolean save_succeeded =
+    g_key_file_save_to_file (priv->metadata_key_file, metadata_filepath, error);
+  g_free (metadata_filepath);
+
+  return save_succeeded;
+}
+
+static gboolean
+add_to_size (EmerCircularFile *self,
+             guint64           delta,
+             GError          **error)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  guint64 new_size = priv->size + delta;
+  guint64 little_endian_size = swap_bytes_64_if_big_endian (new_size);
+  g_key_file_set_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME, SIZE_KEY,
+                         little_endian_size);
+  gboolean save_succeeded = save_metadata_file (self, error);
```

Comment:
For consistency with the function below, you could just make this `if (!save_metadata_file(self, error)) return FALSE;`